src/main/java/com/upserve/uppend/BlockedLongs.java

```diff
@@ -223,39 +212,90 @@ public LongStream values(Long pos) {
             return LongStream.empty();
         }
 
+        long[] longs = valuesArray(pos);
+        return Arrays.stream(longs);
+    }
+
+    public long[] valuesArray(Long pos) {
+
         if (pos < 0 || pos > size()) {
             log.error("Bad position value {} in file {} of size {}", pos, file, size());
+            return new long[]{};
+        }
+
+        // size | -next
+        // prev | -last
+        final long size = readLong(pos);
+
+        if (size < 0) {
+            long nextPos = -size;
+            long[] values = new long[valuesPerBlock];
+            for (int i = 0; i < valuesPerBlock; i++) {
+                values[i] = readLong(pos + 16 + i * 8);
+            }
+
+            long[] additionalValues = valuesArray(nextPos);
+
+            if (additionalValues.length == 0){
+                return values;
+            } else {
+                long[] result = new long[valuesPerBlock + additionalValues.length];
+                System.arraycopy(values,0,result,0, valuesPerBlock);
+                System.arraycopy(additionalValues,0,result,valuesPerBlock, additionalValues.length);
+                return result;
+            }
+        } else if (size > valuesPerBlock) {
+            throw new IllegalStateException("too high num values: expected <= " + valuesPerBlock + ", got " + size);
+        } else if (size == 0) {
+            return new long[]{};
+        } else {
+            int numValues = (int) size;
+            long[] values = new long[numValues];
+            for (int i = 0; i < numValues; i++) {
```

Comment:
Use an `offset` variable here as well?

I have a feeling that this method could be rewritten to make it more clear, and the two loops could probably be combined into one. You could also move the `if(size > valuesPerBlock)` and `if(size == 0)` checks to the top of the ladder and make them separate `if` statements -- no `else` clauses would be needed.

But otoh if you're going for speed, then there's something to be said for putting the most likely case first in the `if` ladder. The value of doing that depends on how often the method gets called, though.

Reply:
You are right - this method is definitely in need of clarification.

I am going to leave the implementation alone for now. I have marked it as deprecated. The next task for Uppend is implementing this method using a spliterator that allows parallel reads.